### ISSUE TITLE
New Action System

### ISSUE THREAD

**Issue by dis446** [dis446/Fallout-Shelter-Text-Based#35]
I don't understand the new action system. Seems like an overly complicated way of doing things. Can't we just call the functions as they are need? What is this add action thing? Could someone please explain this to me quickly?

**Comment by kungtotte** [comment]
Instead of hardcoding every action into a huge if-elif block we do something called dispatching by selecting the appropriate action from a dictionary. We do this by looking up a given action name (in this case it's what the player enters first, like 'skip' or 'show') and then calling the function we have associated with that name.

The add_action function is strictly speaking not required, it's just a convenient way of not repeating ourselves when adding an action.

We separate the logic of each action into a separate function. This makes it easier to document what it does and easier to maintain. It also makes the main game loop shorter and easier to read by having a short way of dispatching actions instead of a big if-elif thing.

We also have all actions stored in a single place (the actions dict) which means we can do things like generate the help text from a docstring in the action functions for example, or store a description like we do now IIRC. That way we would eventually not have to maintain the massive print statement in print_help and we could generate it from the code itself.

If there's any part of the code in particular you don't understand I could explain that in more detail if you want.

**Comment by dis446** [comment]
Thanks. These are some great game practices and I'm glad I'm learning them.

**Comment by sgtlaggy** [comment]
@kungtotte You're completely right about the docstrings (`__doc__`)! I can never remember all the magic functions...

**Comment by dis446** [comment]
Oh okay so any functions that are directly called only when the player chooses to are collated into the actions dictionary and are called appropriately. Great idea. The long if,elif statement was getting a little hard to manage.

**Comment by dis446** [comment]
Quick question. The dispatching works perfectly for printing out the help message and seeing the people/rooms/inventory. However, what about the functions where modifications to the game are being made? Normally, we could just `return game` at the the end of the function and do `game = trade(game)` whenever we call it.
Could we just move the trade function into the game class and make it a method, but still how would we call it when the player input's `trade` with no arguments?

**Comment by dis446** [comment]
Just a thought. We could do a conditional check if 
`a.split()[0] is in ("trade","assign","unnassign"):` and other functions that change things.
and then do `self =  self.actions[action](self, *args)`
and otherwise, just `self.actions[a](self, *args)`

Or maybe modify the actions dictionary to add a boolean argument to track whether or not they change anything in the game.
For example: `self.add_action("quit", action_quit, false)`
and `self.add_action("trade", action_trade, true)` where the last argument tracks whether or not they do anything to the game.

**Comment by dis446** [comment]
So which method should we use. For now, I've implemented the first one, where it checks to see if the first word the player types in is in a list of functions we know take the game as an argument and `return game` at the end.
If the check is True, it calls `self = self.actions[action](self, *args)`
and otherwise, it calls `self.actions[a](self, *args)`

We could also implement the second method.
Or we could move all these functions that modify the game into the Game() class, making them methods and have a separate dispatch system. 

I don't know which solution would be the best.
